content/ember/v3/array-observers.md

```diff
@@ -0,0 +1,345 @@
+---
+id: array-observers
+title: Array Observers
+until: '4.0.0'
+since: 'Upcoming Features'
+---
+
+Array observers are a special type of observer that can be used to synchronously
+react to changes in an `EmberArray`. In general, to refactor away from them, these
+reactions need to be converted from _eager_, _synchronous_ reactions into _lazy_
+reactions that occur when the array in question is _used or accessed_.
+
+For example, let's say that we had a class which wrapped an `EmberArray` and
+converted its contents into strings by calling `toString()` on them. This class
+could be implemented using array observers like so:
+
+```js
+class ToStringArray {
+  constructor(innerArray) {
+    this._inner = innerArray;
+
+    this._content = innerArray.map((value) => value.toString());
+
+    innerArray.addArrayObserver(this, {
+      willChange: '_innerWillChange',
+      didChange: '_innerDidChange',
+    });
+  }
+
+  // no-op
+  _innerWillChange() {}
+
+
+  _innerDidChange(innerArray, changeStart, removeCount, addCount) {
+    if (removeCount) {
+      // if items were removed, remove them
+      this._content.removeAt(changeStart, removeCount);
+    } else {
+      // else, find the new items, convert them, and add them to the array
+      let newItems = innerArray.slice(changeStart, addCount);
+
+      this._content.replace(changeStart, 0, newItems.map((value) => value.toString()));
+    }
+
+    // Let observers/computeds know that the value has changed
+    notifyPropertyChange(this, '[]');
+  }
+
+  objectAt(index) {
+    return this._content.objectAt(index);
+  }
+}
+```
+
+To convert this to no longer use array observers, we could instead convert the
+wrapping to happen when the array is accessed in `objectAt`, using the `@cached`
+decorator from [tracked-toolbox](https://github.com/pzuraq/tracked-toolbox).
+
+```js
+import { cached } from 'tracked-toolbox';
+
+class ToStringArray {
+  constructor(innerArray) {
+    this._inner = innerArray;
+  }
+
+  @cached
+  get _content() {
+    return this._inner.map((value) => value.toString());
+  }
+
+  objectAt(index) {
+    return this._content.objectAt(index);
+  }
+}
+```
+
+This can also be accomplished with native [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy),
+allowing your users to interact with the array using standard array syntax
```

Comment:
I'd write 2 sentences to keep grammar simple.

```suggestion
This can also be accomplished with native [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy).
Your users can interact with the array using standard array syntax
```